File: services/python/shared/ratelimiter.py
import os
import time
import logging
import threading
from typing import Optional

logger = logging.getLogger(__name__)


class TokenBucketRateLimiter:
    def __init__(self, name: str, rate_per_second: int):
        self.name = name
        self.rate = rate_per_second
        self.tokens = float(rate_per_second)
        self.max_tokens = float(rate_per_second)
        self.last_refill = time.monotonic()
        self.total_ops = 0
        self.rejected = 0
        self._lock = threading.Lock()

    def _refill(self):
        now = time.monotonic()
        elapsed = now - self.last_refill
        self.tokens = min(self.max_tokens, self.tokens + elapsed * self.rate)
        self.last_refill = now

    def allow(self) -> bool:
        with self._lock:
            self._refill()
            if self.tokens >= 1:
                self.tokens -= 1
                self.total_ops += 1
                return True
            self.rejected += 1
            return False

    def wait(self, timeout: float = 30.0) -> bool:
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            if self.allow():
                return True
            time.sleep(0.01)
        return False

    def stats(self) -> dict:
        with self._lock:
            return {
                "name": self.name,
                "rate_per_sec": self.rate,
                "total_ops": self.total_ops,
                "rejected": self.rejected,
                "available_tokens": int(self.tokens),
            }


class BatchRateLimiter:
    def __init__(self):
        self._limiters: dict[str, TokenBucketRateLimiter] = {}
        self._lock = threading.Lock()

    def register(self, name: str, rate_per_second: int):
        with self._lock:
            self._limiters[name] = TokenBucketRateLimiter(name, rate_per_second)

    def allow(self, name: str) -> bool:
        with self._lock:
            limiter = self._limiters.get(name)
        if limiter is None:
            return True
        return limiter.allow()

    def wait(self, name: str, timeout: float = 30.0) -> bool:
        with self._lock:
            limiter = self._limiters.get(name)
        if limiter is None:
            return True
        return limiter.wait(timeout)

    def stats(self) -> dict:
        with self._lock:
            return {name: rl.stats() for name, rl in self._limiters.items()}


def default_backfill_limiters() -> BatchRateLimiter:
    brl = BatchRateLimiter()
    brl.register("model-retraining-writes", 100)
    brl.register("lakehouse-etl-inserts", 500)
    brl.register("analytics-backfill", 200)
    brl.register("kyc-batch-verification", 50)
    brl.register("transaction-history-migration", 1000)
    brl.register("fraud-score-recalculation", 150)
    logger.info(f"Backfill rate limiters initialized: {len(brl._limiters)} registered")
    return brl


backfill_limiters = default_backfill_limiters()

File: services/go/shared/ratelimiter.go
package shared

import (
	"context"
	"fmt"
	"sync"
	"time"
)

type RateLimiter struct {
	name       string
	rate       int
	interval   time.Duration
	tokens     chan struct{}
	mu         sync.Mutex
	totalOps   int64
	rejected   int64
	logger     *StructuredLogger
}

func NewRateLimiter(name string, ratePerSecond int, logger *StructuredLogger) *RateLimiter {
	rl := &RateLimiter{
		name:     name,
		rate:     ratePerSecond,
		interval: time.Second / time.Duration(ratePerSecond),
		tokens:   make(chan struct{}, ratePerSecond),
		logger:   logger,
	}

	for i := 0; i < ratePerSecond; i++ {
		rl.tokens <- struct{}{}
	}

	go rl.refill()
	return rl
}

func (rl *RateLimiter) refill() {
	ticker := time.NewTicker(rl.interval)
	defer ticker.Stop()
	for range ticker.C {
		select {
		case rl.tokens <- struct{}{}:
		default:
		}
	}
}

func (rl *RateLimiter) Allow() bool {
	select {
	case <-rl.tokens:
		rl.mu.Lock()
		rl.totalOps++
		rl.mu.Unlock()
		return true
	default:
		rl.mu.Lock()
		rl.rejected++
		rl.mu.Unlock()
		return false
	}
}

func (rl *RateLimiter) Wait(ctx context.Context) error {
	select {
	case <-rl.tokens:
		rl.mu.Lock()
		rl.totalOps++
		rl.mu.Unlock()
		return nil
	case <-ctx.Done():
		rl.mu.Lock()
		rl.rejected++
		rl.mu.Unlock()
		return ctx.Err()
	}
}

func (rl *RateLimiter) Stats() map[string]interface{} {
	rl.mu.Lock()
	defer rl.mu.Unlock()
	return map[string]interface{}{
		"name":         rl.name,
		"rate_per_sec": rl.rate,
		"total_ops":    rl.totalOps,
		"rejected":     rl.rejected,
	}
}

type BatchRateLimiter struct {
	limiters map[string]*RateLimiter
	mu       sync.RWMutex
	logger   *StructuredLogger
}

func NewBatchRateLimiter(logger *StructuredLogger) *BatchRateLimiter {
	return &BatchRateLimiter{
		limiters: make(map[string]*RateLimiter),
		logger:   logger,
	}
}

func (brl *BatchRateLimiter) Register(name string, ratePerSecond int) {
	brl.mu.Lock()
	defer brl.mu.Unlock()
	brl.limiters[name] = NewRateLimiter(name, ratePerSecond, brl.logger)
}

func (brl *BatchRateLimiter) Allow(name string) bool {
	brl.mu.RLock()
	rl, ok := brl.limiters[name]
	brl.mu.RUnlock()
	if !ok {
		return true
	}
	return rl.Allow()
}

func (brl *BatchRateLimiter) Wait(ctx context.Context, name string) error {
	brl.mu.RLock()
	rl, ok := brl.limiters[name]
	brl.mu.RUnlock()
	if !ok {
		return nil
	}
	return rl.Wait(ctx)
}

func (brl *BatchRateLimiter) Stats() map[string]interface{} {
	brl.mu.RLock()
	defer brl.mu.RUnlock()
	stats := make(map[string]interface{})
	for name, rl := range brl.limiters {
		stats[name] = rl.Stats()
	}
	return stats
}

func DefaultBackfillLimiters(logger *StructuredLogger) *BatchRateLimiter {
	brl := NewBatchRateLimiter(logger)
	brl.Register("model-retraining-writes", 100)
	brl.Register("lakehouse-etl-inserts", 500)
	brl.Register("analytics-backfill", 200)
	brl.Register("kyc-batch-verification", 50)
	brl.Register("transaction-history-migration", 1000)
	brl.Register("fraud-score-recalculation", 150)
	logger.Info("backfill rate limiters initialized", map[string]interface{}{
		"limiters": fmt.Sprintf("%d registered", len(brl.limiters)),
	})
	return brl
}
